### PR TITLE
add docker rmi

### PR DESCRIPTION
**Pull request by vieux**
```bash
$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
redis               latest              868be653dea3        5 weeks ago         110.8 MB
redis               latest              868be653dea3        5 weeks ago         110.8 MB
$ docker rmi redis
Untagged: redis:latest
Deleted: 868be653dea3ff6082b043c0f34b95bb180cc82ab14a18d9d6b8e27b7929762c
Deleted: 4380cd9116fa10a0ba96f392d3d8733a198fb5521fb68f2c8b214518f167b1c6
Deleted: cb1becc16e9ff42b18ca229312f1c07e75eab0e499a8fc540079a3546f4135dc
Deleted: bfcd4901e679dd96bf85dda3e232c31b8c700d3a8a0716d266155d49aaa9624b
Deleted: 9bf78464cc9210962d770b88eb6a9f0b74581f8f09cfff8f7756e16fc46e9c0c
Deleted: 9ef278b3f543d5b49cd665cc8feaadeabd118a4be5096b2bd6a9a528e626a64c
Deleted: dbe0ee23d0a635bc6723afe7a6559fa08f6acf5ec18531de3db4aabff8fac43b
Deleted: 239c3ae47786f0a1b85e78d46139aaa9a6405f6a063b64e45f8cd43e2161feea
Deleted: 74f42bf6da0cda3d7d5a3fb2c8724ae321817f0c4df863891bd3ce76c8a54013
Deleted: 2b9e6c467faf34498b938e6f2826c08694ba4914461c1cbf49e2ff62208676e0
Deleted: b2d28145057ecf5a580c60373b5187f3fa2970a3c3086fc8ec42734560e8bfdf
Deleted: 643554a01e20021d05d6dde689ff810f1d97e0e43f7f2ec866d0fbbf452092ca
Deleted: 97047ea9f5f491ebbb0628373e5b89f2ec0b9a3ff8c5d2e8780b700384c5ac17
Deleted: a9be1f2beb10e5f25fadc462298d8d113c39beb32982b06d4bcad85824b764fe
Deleted: b2418d3703c4248e07e98f97933cf61c6a9737d004b947b337eaf949116e846c
Deleted: c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444
Deleted: 30d39e59ffe287f29a41a3f8bd70734afc8728329e3289945cbdc5bbf07cd980
Deleted: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158
Untagged: redis:latest
Deleted: 868be653dea3ff6082b043c0f34b95bb180cc82ab14a18d9d6b8e27b7929762c
Deleted: 4380cd9116fa10a0ba96f392d3d8733a198fb5521fb68f2c8b214518f167b1c6
Deleted: cb1becc16e9ff42b18ca229312f1c07e75eab0e499a8fc540079a3546f4135dc
Deleted: bfcd4901e679dd96bf85dda3e232c31b8c700d3a8a0716d266155d49aaa9624b
Deleted: 9bf78464cc9210962d770b88eb6a9f0b74581f8f09cfff8f7756e16fc46e9c0c
Deleted: 9ef278b3f543d5b49cd665cc8feaadeabd118a4be5096b2bd6a9a528e626a64c
Deleted: dbe0ee23d0a635bc6723afe7a6559fa08f6acf5ec18531de3db4aabff8fac43b
Deleted: 239c3ae47786f0a1b85e78d46139aaa9a6405f6a063b64e45f8cd43e2161feea
Deleted: 74f42bf6da0cda3d7d5a3fb2c8724ae321817f0c4df863891bd3ce76c8a54013
Deleted: 2b9e6c467faf34498b938e6f2826c08694ba4914461c1cbf49e2ff62208676e0
Deleted: b2d28145057ecf5a580c60373b5187f3fa2970a3c3086fc8ec42734560e8bfdf
Deleted: 643554a01e20021d05d6dde689ff810f1d97e0e43f7f2ec866d0fbbf452092ca
Deleted: 97047ea9f5f491ebbb0628373e5b89f2ec0b9a3ff8c5d2e8780b700384c5ac17
Deleted: a9be1f2beb10e5f25fadc462298d8d113c39beb32982b06d4bcad85824b764fe
Deleted: b2418d3703c4248e07e98f97933cf61c6a9737d004b947b337eaf949116e846c
Deleted: c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444
Deleted: 30d39e59ffe287f29a41a3f8bd70734afc8728329e3289945cbdc5bbf07cd980
Deleted: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158
```

### CODE REVIEW

**Comment by vieux** [comment]
ping @aluzzardi

File: api/events_test.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ func (fn *FakeNode) ID() string                            { return "node_id" }
 func (fn *FakeNode) Name() string                          { return "node_name" }
 func (fn *FakeNode) IP() string                            { return "node_ip" }
 func (fn *FakeNode) Addr() string                          { return "node_addr" }
-func (fn *FakeNode) Images() []*cluster.Image              { return nil }
+func (fn *FakeNode) Images(_ string) []*cluster.Image      { return nil }
```
Comment (aluzzardi) [inline api/events_test.go]: What's the difference between Images() and Image()?

Comment (vieux) [inline api/events_test.go]: Sometimes you need only one images (for example for `docker history`)
`cluster.Images(name)[0]` can take much more time than `cluster.Image(name)`

That's why I kept the single image version.
Do you think it's not worth it ?

**Comment by vieux** [comment]
@aluzzardi PTAL

File: api/api.go [hunk 2]
```diff
@@ -330,6 +331,77 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
+
+	size := len(matchedImages)
+	if size == 0 {
+		httpError(w, fmt.Sprintf("No such image %s", name), http.StatusNotFound)
+		return
+	}
+	wf := NewWriteFlusher(w)
+
+	for i, image := range matchedImages {
+		fmt.Println("delete", image)
```
Comment (aluzzardi) [inline api/api.go]: print

File: api/api.go [hunk 3]
```diff
@@ -330,6 +331,77 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
+		req, err := http.NewRequest("DELETE", scheme+"://"+image.Node.Addr()+"/images/"+name, nil)
```
Comment (aluzzardi) [inline api/api.go]: Why http request rather than relying on dockerclient?

File: api/api.go [hunk 4]
```diff
@@ -330,6 +331,77 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
+		if err != nil {
+			if size == 1 {
+				httpError(w, err.Error(), http.StatusInternalServerError)
+			}
+			continue
+		}
+		resp, err := client.Do(req)
```
Comment (aluzzardi) [inline api/api.go]: Unnecessary with dockerclient

File: api/api.go [hunk 5]
```diff
@@ -330,6 +331,77 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
+		if err != nil {
+			if size == 1 {
+				httpError(w, err.Error(), http.StatusInternalServerError)
+			}
+			continue
+		}
+		data, err := ioutil.ReadAll(resp.Body)
```
Comment (aluzzardi) [inline api/api.go]: Same here

File: api/api.go [hunk 6]
```diff
@@ -330,6 +331,77 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
+		httpError(w, fmt.Sprintf("No such image %s", name), http.StatusNotFound)
+		return
+	}
+	wf := NewWriteFlusher(w)
+
+	for i, image := range matchedImages {
+		fmt.Println("delete", image)
+		req, err := http.NewRequest("DELETE", scheme+"://"+image.Node.Addr()+"/images/"+name, nil)
+		if err != nil {
+			if size == 1 {
```
Comment (aluzzardi) [inline api/api.go]: what happens when `size > 1`?

**Comment by vieux** [comment]
@aluzzardi PTAL

File: api/api.go [hunk 7]
```diff
@@ -330,6 +330,46 @@ func postContainersExec(c *context, w http.ResponseWriter, r *http.Request) {
...
+
+	if len(matchedImages) == 0 {
+		httpError(w, fmt.Sprintf("No such image %s", name), http.StatusNotFound)
+		return
+	}
+
+	out := []*dockerclient.ImageDelete{}
+	errs := []string{}
+	for _, image := range matchedImages {
+		content, err := image.Node.DockerClient().RemoveImage(name)
```
Comment (aluzzardi) [inline api/api.go]: On second thought, I would add the remove image functionality directly to the Node so we don't leak the docker client.

Comment (vieux) [inline api/api.go]: @aluzzardi What about `cluster.RemoveImage(image)` right now the node has no actions, only "getters" and I think it's a good thing

Comment (aluzzardi) [inline api/api.go]: +1 on that!

Also, what about having `cluster.RemoveImage(image)`, as you mentioned, which will do the same loop as here, but in the end will call `node.removeImage` on each node or something like that, that actually calls dockerclient, similar to `cluster.CreateContainer`/`node.create` and `cluster.RemoveContainer`/`node.destroy`?

Comment (vieux) [inline api/api.go]: will do!

**Comment by vieux** [comment]
@aluzzardi PTAAL

**Comment by vieux** [comment]
ping @abronan

**Comment by abronan** [comment]
LGTM

**Pull request merged**